zcash_client_backend/src/welding_rig.rs

```diff
@@ -68,38 +66,111 @@ fn scan_output<P: consensus::Parameters>(
         // - Notes sent from one account to itself.
         let is_change = spent_from_accounts.contains(&account);
 
+        let witness = IncrementalWitness::from_tree(tree);
+        let nf = vk.nf(&note, &witness);
+
         return Some(WalletShieldedOutput {
             index,
             cmu,
             epk,
-            account: *account,
+            account: **account,
             note,
             to,
             is_change,
-            witness: IncrementalWitness::from_tree(tree),
+            witness,
+            nf,
         });
     }
+
     None
 }
 
-/// Scans a [`CompactBlock`] with a set of [`ExtendedFullViewingKey`]s.
+/// A key that can be used to perform trial decryption and nullifier
+/// computation for a Sapling [`CompactOutput`]
+///
+/// [`CompactOutput`]: crate::proto::compact_formats::CompactOutput
+pub trait ScanningKey {
```

Comment:
minor observation: Both implementations of this trait call `try_sapling_compact_note_decryption` would it simplify things to either a) not rename the function for this trait or b) refactor `note_encryption` to change the name of both functions to `try_decryption`

Reply:
The implementation of `try_decryption` varies in terms of what part of `self` is passed to `try_sapling_compact_note_decryption`. This is the other purpose of this trait.